tech_reports/LLMs/llms.md

```diff
@@ -52,12 +52,199 @@ Authors:
     - which dims are parallelized
 ### 2.5 MLP
 ### 2.6 Decoder
+<div align="center">
+<img src="decoder.png" alt="Decoder Diagram" title="Decoder Title" width="350" height="400">
+</div> <br>
+If the components explained in previous sections (MLP, Attention, RMSNorm) are implemented, bringing up the decoder should be relatively straightforward. According to the diagram (based on the Llama3.1 example), the components are stacked sequentially during the forward pass. Only thing to worry about is whether addition of MLP and Attention outputs should be stored in L1 or in DRAM. <br><br> Decode forward pass implementation below follows diagram above and has nothing that is not already explained in previous sections. Also, it's crucial to deallocate tensors after their usage to optimize memory. However, in this explanation, tensor deallocation has been omitted for clarity and to keep the code as straightforward as possible.
```

Comment:
If it is crucial to deallocate tensors, then the example should show the deallocations.

Reply:
I re-worded this a bit as it's not crucial to have deallocates, but it's recommended for tight memory runs.